### PR TITLE
Warn if optimistic state is updated outside of a transition 

### PR DESCRIPTION
**Pull request by acdlite**
### Based on #27453 

If optimistic state is updated, and there's no startTransition on the stack, there are two likely scenarios.

One possibility is that the optimistic update is triggered by a regular event handler (e.g. `onSubmit`) instead of an action. This is a mistake and we will warn.

The other possibility is the optimistic update is inside an async action, but after an `await`. In this case, we can make it "just work" by associating the optimistic update with the pending async action.

Technically it's possible that the optimistic update is unrelated to the pending action, but we don't have a way of knowing this for sure because browsers currently do not provide a way to track async scope. (The AsyncContext proposal, if it lands, will solve this in the future.) However, this is no different than the problem of unrelated transitions being grouped together — it's not wrong per se, but it's not ideal.

Once AsyncContext starts landing in browsers, we will provide better warnings in development for these cases.

### CODE REVIEW

**Review by sebmarkbage — APPROVED** [review]

**Comment by sophiebits** [comment]
A warning is good but does this mean if you call set outside a transition then that update will stick around forever? Can we not make it so that it's never applied / immediately reverted? I worry people might build the wrong mental model otherwise.

**Comment by acdlite** [comment]
> A warning is good but does this mean if you call set outside a transition then that update will stick around forever? Can we not make it so that it's never applied / immediately reverted? I worry people might build the wrong mental model otherwise.

Yeah it does get immediately reverted... now that https://github.com/facebook/react/pull/27453 landed :D That's how I noticed it

https://github.com/facebook/react/blob/ba826abbddf2b8d004e1272e535a69002160dd3c/packages/react-reconciler/src/__tests__/ReactAsyncActions-test.js#L1223-L1224

**Comment by acdlite** [comment]
there is a failing test related to when `enableAsyncActions` is disabled but it's caused by something else, I'll fix after this meeting

File: packages/react-reconciler/src/__tests__/ReactAsyncActions-test.js [hunk 1]
```diff
@@ -1175,4 +1175,52 @@ describe('ReactAsyncActions', () => {
...
+      const [, _startTransition] = useTransition();
+      const [text, _setText] = useState('A');
+      const [loadingProgress, _setLoadingProgress] = useOptimistic(0);
+      startTransition = _startTransition;
+      setText = _setText;
+      setLoadingProgress = _setLoadingProgress;
+
+      return (
+        <>
+          {loadingProgress !== 0 ? (
```
Comment (sophiebits) [inline packages/react-reconciler/src/__tests__/ReactAsyncActions-test.js:1194]: Maybe make this unconditional so it’s easier to witness that the state gets set back? Right now the test logs sorta make it look like it gets stuck at 25.

Comment (sophiebits) [reply packages/react-reconciler/src/__tests__/ReactAsyncActions-test.js:1194]: Wait never mind I'm silly. The toMatchRenderedOutput is plenty.

**Pull request merged**